tensorflow/compiler/xla/service/gpu/gemm_algorithm_picker.cc

```diff
@@ -219,49 +203,222 @@ static StatusOr<absl::optional<se::blas::AlgorithmType>> DoUncachedGemmAutotune(
 static StatusOr<absl::optional<se::blas::AlgorithmType>> DoGemmAutotune(
     const HloInstruction* instr, const GemmBackendConfig& gemm_config,
     se::DeviceMemoryAllocator* allocator, se::Stream* stream) {
+  VLOG(3) << "Starting autotune of GemmThunk " << instr->ToString();
   const HloInstruction* lhs = instr->operand(0);
   const HloInstruction* rhs = instr->operand(1);
 
   // Don't run autotuning concurrently on the same GPU.
   tensorflow::mutex_lock gpu_lock = LockGpu(stream->parent());
+  const HloModuleConfig& hlo_module_config = instr->GetModule()->config();
+  const bool crash_on_checking_failure =
+      hlo_module_config.debug_options()
+          .xla_gpu_crash_on_verification_failures();
+  const int32_t cublas_autotune_level =
+      hlo_module_config.debug_options().xla_gpu_autotune_level();
+  const bool init_cublas_data = cublas_autotune_level >= 2;
 
-  GemmCacheKey key =
-      std::make_tuple(stream->parent(), lhs->shape(), rhs->shape(),
-                      instr->shape(), gemm_config.SerializeAsString());
+  se::RedzoneAllocator input_output_allocator(
+      stream, allocator,
+      PtxOptsFromDebugOptions(hlo_module_config.debug_options()),
+      /*memory_limit=*/std::numeric_limits<int64_t>::max());
+  int64_t rng_state = 0;
+  auto get_initialized_buffer =
+      [&](const HloInstruction* op) -> StatusOr<se::DeviceMemoryBase> {
+    TF_ASSIGN_OR_RETURN(se::DeviceMemoryBase buffer,
+                        input_output_allocator.AllocateBytes(
+                            ShapeUtil::ByteSizeOf(op->shape())));
+    if (init_cublas_data) {
+      InitializeBuffer(stream, op->shape().element_type(), &rng_state, buffer);
+    }
+    return buffer;
+  };
 
-  tensorflow::mutex_lock cache_lock(autotune_cache_mu);
-  auto it = autotune_cache.find(key);
-  int64_t autotuning_requests = cache_hits + cache_misses;
-  if (autotuning_requests && autotuning_requests % 10 == 0) {
-    VLOG(2) << "Autotuning cache hits/(hits + misses): " << cache_hits << "/"
-            << autotuning_requests;
-  }
+  TF_ASSIGN_OR_RETURN(se::DeviceMemoryBase lhs_buffer,
+                      get_initialized_buffer(instr->operand(0)));
+  TF_ASSIGN_OR_RETURN(se::DeviceMemoryBase rhs_buffer,
+                      get_initialized_buffer(instr->operand(1)));
+  TF_ASSIGN_OR_RETURN(se::DeviceMemoryBase output_buffer,
+                      get_initialized_buffer(instr));
+  TF_ASSIGN_OR_RETURN(se::DeviceMemoryBase reference_result_buffer,
+                      get_initialized_buffer(instr));
+
+  int64_t batch_size = gemm_config.batch_size();
+  GpuGemmConfig config = GetGpuGemmConfig(instr);
+  const Shape& output_shape = config.output_shape;
+  absl::flat_hash_set<PrimitiveType> enabled_types = {F16, F32, F64, C64, C128};
+
+  if (config.use_cublaslt && stream->parent()->SupportsBlasPlans() &&
+      enabled_types.contains(output_shape.element_type())) {
```

Comment:
The nesting level gets really high at this point, and it's a bit hard to follow, could we move the logic for cublasLT autotuning into a separate function?